recirq/quantum_chess/swap_updater.py

```diff
@@ -0,0 +1,140 @@
+      swap_factory: the factory used to produce operations representing a swap
+        of two qubits
+    """
+    def __init__(self,
+                 circuit: cirq.Circuit,
+                 device_qubits: Optional[Iterable[cirq.GridQubit]],
+                 initial_mapping: Dict[cirq.Qid, cirq.GridQubit] = {},
+                 swap_factory: Callable[
+                     [cirq.Qid, cirq.Qid],
+                     List[cirq.Operation]] = generate_decomposed_swap):
+        self.device_qubits = device_qubits
+        self.dlists = mcpe.DependencyLists(circuit)
+        self.mapping = mcpe.QubitMapping(initial_mapping)
+        self.swap_factory = swap_factory
+
+    def generate_candidate_swaps(
+        self, gates: Iterable[cirq.Operation]
+    ) -> Generator[Tuple[cirq.GridQubit, cirq.GridQubit], None, None]:
+        """Generates the candidate SWAPs that would have a positive effect on at
+        least one of the given physical gates.
+
+        Args:
+          gates: the list of gates to consider which operate on GridQubits
+        """
+        for gate in gates:
+            for gate_q in gate.qubits:
+                yield from (
+                    (gate_q, swap_q)
+                    for swap_q in gate_q.neighbors(self.device_qubits)
+                    if mcpe.effect_of_swap((gate_q, swap_q), gate.qubits) > 0)
+
+    def update_iteration(self) -> Generator[cirq.Operation, None, None]:
+        """Runs one iteration of the swap update algorithm and updates internal
+        state about the original circuit.
+
+        Returns:
+          the operations on GridQubits in the final updated circuit generated by
+          this iteration
+        """
+        # Handle the already-satisfied active gates.
+        # Those can be immediately added into the final circuit.
+        active_physical_gates = []
+        for gate in self.dlists.active_gates():
+            physical_gate = gate.transform_qubits(self.mapping.physical)
+            if _satisfies_adjacency(physical_gate):
+                # physical_gate is ready to be popped off the dependecy lists
+                # and added to the final circuit.
+                for q in gate.qubits:
+                    self.dlists.pop_front(q)
+                yield physical_gate
+            else:
+                # physical_gate needs to be fixed up with some swaps.
+                active_physical_gates.append(physical_gate)
+
+        # If all the active gates in this pass were added to the final circuit,
+        # then we have nothing left to do until we make another pass and get the
+        # newly active gates.
+        if not active_physical_gates:
+            return
+
+        candidates = set(self.generate_candidate_swaps(active_physical_gates))
+        # This could happen on irregular topologies or when some device
+        # qubits are disallowed.
+        assert candidates, 'no swaps found that will improve the circuit'
```

Comment:
Will this also happen if the mapping is already optimal?  I am not sure whether asserting here is the correct thing to do.

Reply:
if the mapping were already optimal, we'd always fall out from the `if not active_physical_gates` condition just before this (I will add some test coverage for already-optimal circuits regardless).